### DISCUSSION TITLE
some discrepancies in .env and README.md

### DISCUSSION THREAD

**Discussion by anfedoro** [herlesupreeth/docker_open5gs#299] [General]
Hello.. I have successfuly installed the docker_open5sg for 4G LTE network. I use srsenb and srsue with mqz RF simulation.
what I found that there are UE related settings (IMSI, OP/OPC, KI) are different in .env and README.md guidance for installation.

Am I right  that I have to change these settings in .env and rebuild (or just recompose?) srsue_mqz. Also looks like I have to update the srsue_mqz.cfg file to use UE_OPC instead of UE_OP.
I suppose AuC settings for OP or OPc have to be concistent with UE SIM configuration and I found that pyHSS  OAM API doesnt understand "op' parameter.. only "opc"

please advice what should/ shouldnt be changed ?

thanks

**Comment by herlesupreeth** [comment]
> what I found that there are UE related settings (IMSI, OP/OPC, KI) are different in .env and README.md

That is right. Just ensure to match the IMSI. OP/OPc and Ki are same

> Am I right that I have to change these settings in .env and rebuild (or just recompose)?

no need to rebuild. Just re-deploy

> Also looks like I have to update the srsue_mqz.cfg file to use UE_OPC instead of UE_OP.

No need to change since OPc is derived from Ki and OP

**Reply by anfedoro** [reply]
Sorry,  am I understand right that we configure
 "ki": "8baf473f2f8fd09487cccbd7097c6862",
  "opc": "8E27B6AF0E692E750F32667A3B14605D"
  in both open5gs HSS and pyHSS AuC
  while UE keep with UE1_OP=11111111111111111111111111111111?

How to ensure such OP  and Ki gives configured OPC ? 

and if so (OPc is result of OP and KI) I even dont need to redeploy.. only fill HSS profiles with correct IMSI, KI, OPC respective to UE

how to add one more UE  and how to test VoLTE registration/call from such UE ? is it possble at all ?

**Reply by anfedoro** [reply]
also.. shouldn's here sh_profile refer to default_sh_user_data.xml ?
{
     "imsi": "001010123456790",
    "msisdn": "9076543210",
    "sh_profile": "string",
    "scscf_peer": "scscf.ims.mnc001.mcc001.3gppnetwork.org",
    "msisdn_list": "[9076543210]",
    "ifc_path": "default_ifc.xml",
    "scscf": "sip:scscf.ims.mnc001.mcc001.3gppnetwork.org:6060",
    "scscf_realm": "ims.mnc001.mcc001.3gppnetwork.org"
}

**Comment by herlesupreeth** [comment]
> Sorry, am I understand right that we configure
> "ki": "8baf473f2f8fd09487cccbd7097c6862",
> "opc": "8E27B6AF0E692E750F32667A3B14605D"
> in both open5gs HSS and pyHSS AuC
> while UE keep with UE1_OP=11111111111111111111111111111111?

yes

> How to ensure such OP and Ki gives configured OPC ?

You can choose any value for Ki and OP and use the algorithm to calculate OPc. As long as you keep Ki and OP same you get same OPc

> how to add one more UE and how to test VoLTE registration/call from such UE ? is it possble at all ?

If you are using ZMQ then its not possible to test VoLTE registration/call since there is no application that glues together the srsRAN UE stack to IMS stack to simulate ideal VoLTE scenarios. However, you can run IMS stack and srsRAN UE to simulate a VoIP registration/call.

> also.. shouldn's here sh_profile refer to default_sh_user_data.xml ?
> {
> "imsi": "001010123456790",
> "msisdn": "9076543210",
> "sh_profile": "string",
> "scscf_peer": "scscf.ims.mnc001.mcc001.3gppnetwork.org",
> "msisdn_list": "[9076543210]",
> "ifc_path": "default_ifc.xml",
> "scscf": "sip:scscf.ims.mnc001.mcc001.3gppnetwork.org:6060",
> "scscf_realm": "ims.mnc001.mcc001.3gppnetwork.org"
> }

yes, its a typo in the README. Will fix it

**Reply by anfedoro** [reply]
Good.. thanks

Few more notes:
 - finally I have managed to run the setup, means UE get attached and I can ping UPF and seems like the mobile connectivity works, but... once UE getting IP from UPF it still keep default route via eth0. Not sure how to force UPF to push new default route via tun_srsue? I have manually updated original default route with metric 100 and added new one via mobile connection with metric 0. Then all traffic (not only  targeted to UPF) is going  there.
 - then I realised that I can ping or traceroute any host in internet, but when I tried to create more traffic using wget, nothing works. tcpdump at UPF shows that all stuck after tcp handshacking, but there is no http get after syn, ack.. only TCP keepalives from UE after some timeot until RST. Curious that icmp is passing thorugh, also DNS queries are. But no any TCP traffic can.
 Not shue how to raubleshoot and what to change.

Also, would you advice how to use IMS stack to simulate VoLTE registration from srsUE ? which app should I use ?
I thought I can change deployment yamp a bit to have second UE and assgn default APN to ims one. But what to do then ?
even if I install some IMS app in the srsUE, e.g. such as SIPp, how to manage it to use UE SIM credentials to authenticate in IMS ?

Sorry for many questions :) and thanks in advance.

**Comment by herlesupreeth** [comment]
> finally I have managed to run the setup, means UE get attached and I can ping UPF and seems like the mobile connectivity works, but... once UE getting IP from UPF it still keep default route via eth0. Not sure how to force UPF to push new default route via tun_srsue?

Its normal to keep the default route via eth0. If you remove that it could mess up the connectivity to container to docker network. If you want to force traffic to go through tun_srsue then the application you use should bind to that IP/interface. For example, iperf has -B option to bind to an IP and ping also has an -I option to bind to an interface

> then I realised that I can ping or traceroute any host in internet, but when I tried to create more traffic using wget, nothing works. tcpdump at UPF shows that all stuck after tcp handshacking, but there is no http get after syn, ack.. only TCP keepalives from UE after some timeot until RST.

docker network iptables could be filtering those packet. I would avoid changing default route in the container.

> Also, would you advice how to use IMS stack to simulate VoLTE registration from srsUE ? which app should I use ?

I dont have experience here as I always used a COTS UE to test VoLTE. But I have seen other suggesting IMSDroid or sipp to simulate a IMS registration

> I thought I can change deployment yamp a bit to have second UE

This should be possible, just need to give a different IP and container name than the first UE.

> even if I install some IMS app in the srsUE, e.g. such as SIPp, how to manage it to use UE SIM credentials to authenticate in IMS ?

I am not sure whether sipp support IMS AKA Digest based authentication. I think it supports Digest MD5.

**Reply by anfedoro** [reply]
Thanks for you response.
I found that Sipp can support Digest/AKA ("algorithm="AKAv1-MD5"", as specified by 3GPP for IMS). so far this is a target client ot simulate VoLTE calls.

I suppose to test Volte I may configure srsUE with default IMS APN and then test it seprately from data one.

As for routing I didn't remove eth0 default route.. i just changet its metric to 100. And then, when I add new default via tun_srsue with metric 0 it get priority and forward all traffic (except directly connected 172.22.x.x network) through mobile connection.
Actually I didint get how dockers iptables may impact  of the TCP trafic forwarding ? why wget directly from UPF does work.. but same coming from GTP does not ? I suppose UPF does NAT for such traffic once it decapsulated from GTP and to be forwarded further. Interesting that TCP SYN os successfully passing trough and geting response.. but when, due some reason it doesnt send HTTP get, but continue with keepalives until RST.
![telegram-cloud-photo-size-2-5217537094954639756-y](https://github.com/herlesupreeth/docker_open5gs/assets/40258722/8b9ec362-74e3-4c40-b129-458291d6bf55)
192.168.17.4 there is the docker host IP.. and dst IP is amazons one I am trying to make wget from.
